model/scikit/dffml_model_scikit/scikit_base.py

```diff
@@ -119,6 +121,94 @@ def _filename(self):
             yield repo
 
 
+class ScikitContextUnsprvised(ScikitContext):
+    def __init__(self, parent):
+        super().__init__(parent)
+
+    def _feature_predict_hash(self):
+        params = "".join(
+            [
+                "{}{}".format(k, v)
+                for k, v in self.parent.config._asdict().items()
+                if k not in ["directory", "features"]
+            ]
+        )
+        return hashlib.sha384(
+            "".join([params] + self.features).encode()
+        ).hexdigest()
+
+    async def __aenter__(self):
+        if os.path.isfile(self._filename()):
+            self.clf = joblib.load(self._filename())
+        else:
+            config = self.parent.config._asdict()
+            del config["directory"]
+            del config["features"]
+            self.clf = self.parent.SCIKIT_MODEL(**config)
+        return self
+
+    async def train(self, sources: Sources):
+        data = []
+        async for repo in sources.with_features(self.features):
+            feature_data = repo.features(self.features)
+            data.append(feature_data)
+        df = pd.DataFrame(data)
+        xdata = np.array(df)
+        self.logger.info("Number of input repos: {}".format(len(xdata)))
+        self.clf.fit(xdata)
+        joblib.dump(self.clf, self._filename())
+
+    async def accuracy(self, sources: Sources) -> Accuracy:
+        if not os.path.isfile(self._filename()):
+            raise ModelNotTrained("Train model before assessing for accuracy.")
+        data = []
+        async for repo in sources.with_features(self.features):
+            feature_data = repo.features(self.features)
+            data.append(feature_data)
+        df = pd.DataFrame(data)
+        xdata = np.array(df)
+        self.logger.debug("Number of input repos: {}".format(len(xdata)))
+        estimator_type = getattr(self.clf, "_estimator_type")
+        if estimator_type is "clusterer":
+            # measures 'quality' of model
+            if hasattr(self.clf, 'predict'):
+                ydata = self.clf.predict(xdata)
+            else:
+                ydata = self.clf.labels_.astype(np.int)
+            self.confidence = silhouette_score(xdata, ydata)
+        self.logger.debug("Model Accuracy: {}".format(self.confidence))
+        return self.confidence
+
+    async def predict(
+        self, repos: AsyncIterator[Repo]
+    ) -> AsyncIterator[Tuple[Repo, Any, float]]:
+        if not os.path.isfile(self._filename()):
+            raise ModelNotTrained("Train model before prediction.")
+
+        estimator_type = getattr(self.clf, "_estimator_type")
+        if estimator_type is "clusterer":
+            if hasattr(self.clf, "predict"):
+                # inductive clusterer
+                predictor = self.clf.predict
+            else:
+                # transductive clusterer
+                labels = [
+                    (yield label) for label in self.clf.labels_.astype(np.int)
```

Comment:
so, here shouldn't the model be retrained? or is it some form of online learning?

Reply:
[Transductive](https://scikit-learn.org/stable/glossary.html#term-transductive) methods (e.g. [OPTICS,](https://scikit-learn.org/stable/modules/generated/sklearn.cluster.OPTICS.html#sklearn.cluster.OPTICS) [SpectralClustering](https://scikit-learn.org/stable/modules/generated/sklearn.cluster.SpectralClustering.html#sklearn.cluster.SpectralClustering) ) can only make predictions on data which was used to train them. That's why for such methods the output corresponds to the predictions made on training data (calling `fit`  trains as well as predicts output as `labels_` for training data, also they don't have `predict` method ). This will probably have to be documented, informing the user about the methods which need to have prediction data same as training data.

Reply:
mm.. so SpetralClustering is O(n^3) and OPTICS is O(n^2) , I would say both implementations in its current form are irrelevant for practical analysis when the dataset is not *tiny*.

I would go with K-means and it that does not work by default on the data, some feature engineering (changing coordinate system for eg.) can help with clustering the data.. so is there a *real* need for adding this type of algorithms today?

Reply:
You are right. Both of these are computationally expensive and K-means remain unbeatable in most of the realistic scenarios. Considering this, I don't think there is any real need other than giving choice to the users to select what works best for their use case. So, should I drop them for now?

Reply:
Lets warn with `self.logger.critical` when we hit this path. Mention that prediction must be done with the same data source that the model was trained with. Also give a warning about what Rahul said about needing to keep the dataset small.